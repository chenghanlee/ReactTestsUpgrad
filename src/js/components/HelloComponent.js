import React, { Component } from 'react';

import '../../css/tachyons.min.css';

export default class Hello extends Component {
    constructor(props) {
        super(props);
        this.state = {
            buttonClicked: false,
        };
    }

    handleOnClick = e => {
        e.preventDefault();

        const buttonClickedState = this.state.buttonClicked;
        this.setState({ buttonClicked: !buttonClickedState });
    };

    render() {
        const buttonColor = this.state.buttonClicked ? 'bg-red' : 'bg-dark-blue';
        const buttonClassNames = `link tc ph3 pv2 db bg-animate ${buttonColor} white f6 b br1`;

        return (
            <div className="w-100 sans-serif">
                <section className="tc pa3 pa5-ns">
                    <article className="hide-child relative ba b--black-20 mw5 center">
                        <img src="https://i.imgur.com/VZgxXSz.jpg" className="db" alt="Photo of the Terminator" />
                        <div className="pt2 bt b--black-20">
                            <a className="f6 db link dark-blue hover-blue" href="#">
                                Arnold S.
                            </a>
                            <p className="f6 gray mv1 pb3">1 mutual friend</p>
                            <a id="js-button" className={buttonClassNames} href="#" onClick={this.handleOnClick}>
                                Add Friend
                            </a>
                        </div>
                    </article>
                </section>
            </div>
        );
    }
}
