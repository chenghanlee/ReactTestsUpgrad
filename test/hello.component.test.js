import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import Adapter from 'enzyme-adapter-react-16';
import HelloComponent from '../src/js/components/HelloComponent';

Enzyme.configure({ adapter: new Adapter() });

test('Hello Component Renders Correctly', () => {
    const hello = shallow(<HelloComponent />);

    // tests that the render HTML contains a text called "Add Friend"
    expect(hello.text()).toContain('Add Friend');
});

test('Test That The Component Renders Correctly Using Snapshots', () => {
    // Snapshot testing explanined:
    // https://medium.com/@luisvieira_gmr/snapshot-testing-react-components-with-jest-3455d73932a4
    const hello = shallow(<HelloComponent />);
    expect(shallowToJson(hello)).toMatchSnapshot();
});

test('Component Handles onClick Correctly', () => {
    const buttonId = '#js-button';
    const hello = shallow(<HelloComponent />);
    const mockedEvent = { preventDefault: () => {} };

    hello.find(buttonId).simulate('click', mockedEvent);
    expect(hello.find(buttonId).hasClass('bg-red')).toEqual(true);
});

test('Component Initialize Its State Correctly', () => {
    const buttonId = '#js-button';
    const hello = shallow(<HelloComponent />);
    const mockedEvent = { preventDefault: () => {} };

    // buttonState is not clicked. Therefore, it is false
    expect(hello.state().buttonClicked).toEqual(false);
});

test('Component Handles State Changes Correctly', () => {
    const buttonId = '#js-button';
    const hello = shallow(<HelloComponent />);
    const mockedEvent = { preventDefault: () => {} };

    // Simulate clicking the button
    hello.find(buttonId).simulate('click', mockedEvent);
    // buttonClicked state should now be true
    expect(hello.state().buttonClicked).toEqual(true);

    // Simulate clicking the button once again to undo the previous click
    hello.find(buttonId).simulate('click', mockedEvent);
    // buttonClicked state should now be false
    expect(hello.state().buttonClicked).toEqual(false);
});
