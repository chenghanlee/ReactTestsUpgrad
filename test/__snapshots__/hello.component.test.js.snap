// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test That The Component Renders Correctly Using Snapshots 1`] = `
<div
  className="w-100 sans-serif"
>
  <section
    className="tc pa3 pa5-ns"
  >
    <article
      className="hide-child relative ba b--black-20 mw5 center"
    >
      <img
        alt="Photo of the Terminator"
        className="db"
        src="https://i.imgur.com/VZgxXSz.jpg"
      />
      <div
        className="pt2 bt b--black-20"
      >
        <a
          className="f6 db link dark-blue hover-blue"
          href="#"
        >
          Arnold S.
        </a>
        <p
          className="f6 gray mv1 pb3"
        >
          1 mutual friend
        </p>
        <a
          className="link tc ph3 pv2 db bg-animate bg-dark-blue white f6 b br1"
          href="#"
          id="js-button"
          onClick={[Function]}
        >
          Add Friend
        </a>
      </div>
    </article>
  </section>
</div>
`;
